lib/jxl/enc_ac_strategy.cc

```diff
@@ -481,77 +400,98 @@ float EstimateEntropy(const AcStrategy& acs, size_t x, size_t y,
+  for (auto tx : kTransforms8x8) {
+    AcStrategy acs = AcStrategy::FromRawStrategy(tx.type);
+    float entropy =
+      EstimateEntropy(acs, x, y, config,
+		      cmap_factors, block, scratch_space, quantized);
+    entropy = tx.entropy_add + tx.entropy_mul * entropy;
+    if (entropy < best) {
+      best_tx = tx.type;
+      best = entropy;
     }
   }
-  // Nothing changed.
-  if (best == num_candidates) return;
-  AcStrategy acs = AcStrategy::FromRawStrategy(candidates[best]);
-  size_t idx = 0;
-  for (size_t y = 0; y < current_acs.covered_blocks_y();
-       y += acs.covered_blocks_y()) {
-    for (size_t x = 0; x < current_acs.covered_blocks_x();
-         x += acs.covered_blocks_x()) {
-      ac_strategy->Set(bx + x, by + y, candidates[best]);
-      for (size_t iy = y; iy < y + acs.covered_blocks_y(); iy++) {
-        for (size_t ix = x; ix < x + acs.covered_blocks_x(); ix++) {
-          entropy_estimate[iy * 8 + ix] = ee_val[best][idx];
-        }
+  *entropy_out = best;
+  return best_tx;
+}
+
+// bx, by addresses the 64x64 block at 8x8 subresolution
+// cx, cy addresses the left, upper 8x8 block position of the candidate
+// transform.
+void TryMergeToACSCandidate(AcStrategy::Type acs_raw,
+			    size_t bx, size_t by,
+			    size_t cx, size_t cy,
+			    const ACSConfig& config,
+			    const float* JXL_RESTRICT cmap_factors,
+			    AcStrategyImage* JXL_RESTRICT ac_strategy,
+			    const float entropy_mul,
+			    const uint8_t candidate_priority,
+			    uint8_t *priority,
+			    float* JXL_RESTRICT entropy_estimate, float* block,
+			    float* scratch_space, uint32_t* quantized) {
+  AcStrategy acs = AcStrategy::FromRawStrategy(acs_raw);
+  float entropy_current = 0;
+  for (size_t iy = 0; iy < acs.covered_blocks_y(); ++iy) {
+    for (size_t ix = 0; ix < acs.covered_blocks_x(); ++ix) {
+      if (priority[(cy + iy) * 8 + (cx + ix)] >= candidate_priority) {
+	return;
       }
-      idx++;
+      entropy_current += entropy_estimate[(cy + iy) * 8 + (cx + ix)];
+    }
+  }
+  float entropy_candidate = entropy_mul *
+    EstimateEntropy(acs, (bx + cx) * 8, (by + cy) * 8, config,
+		    cmap_factors, block, scratch_space, quantized);
+
+  // printf("hipsu %g %g\n", entropy_candidate, entropy_current);
+  if (entropy_candidate >= entropy_current) return;
+
+  // Accept the candidate.
+  for (size_t iy = 0; iy < acs.covered_blocks_y(); iy++) {
+    for (size_t ix = 0; ix < acs.covered_blocks_x(); ix++) {
+      entropy_estimate[(cy + iy) * 8 + cx + ix] = 0;
+      priority[(cy + iy) * 8 + cx + ix] = candidate_priority;
+      ac_strategy->Set(bx + cx + ix, by + cy + iy, static_cast<AcStrategy::Type>(0));
     }
   }
+  ac_strategy->Set(bx + cx, by + cy, acs_raw);
+  entropy_estimate[cy * 8 + cx] = entropy_candidate;
 }
 
 void ProcessRectACS(PassesEncoderState* JXL_RESTRICT enc_state,
                     const ACSConfig& config, float* entropy_adjust,
                     const Rect& rect) {
```

Comment:
Naive question: why is this done on 64x64 rects, and not group-sized rects? To stay aligned with the chroma from luma map? (but that's not really needed, is it?) For speed?

Doing it on 256x256 rects can give more opportunities to merge blocks into larger ones, and also to make more effective use of the larger DCTs — e.g. 64x64 would have to be lucky to naturally align, but it's more likely that (a non-naturally aligned) one could be found through bottom-up merging. Same with 32x32 that crosses 64x64 boundaries, etc.

Reply:
Oh wait, this algorithm still produces only naturally aligned acs maps, right? No floating blocks?

Reply:
Our encoder currently does very bad things for blocks that go out of 64x64 regions...

Reply:
OK, not for this pull request, but why?

Reply:
IIRC it has to do with CfL, but I don't recall exactly what's wrong there

Reply:
yes, CfL is 64x64 and making them diverge creates additional considerations that require more care than I can afford now